batch/batch/front_end/front_end.py

```diff
@@ -733,13 +742,10 @@ async def create_jobs(request, userdata):
 
                 inst_coll_configs: InstanceCollectionConfigs = app['inst_coll_configs']
 
-                result, exc = inst_coll_configs.select_inst_coll(
+                result = inst_coll_configs.select_inst_coll(
```

Comment:
So, I stared at this code for a bit and then at the InstanceCollectionManager and I found it a bit hard to follow. Some concrete criticisms:
- I dislike repeated conditionals, e.g. we branch on `machine_type is None` twice.
- There's too many different, actively-used variables and fields for a single concept. Consider `cpu`:  the field `resources['cpu']`, the variable `req_cores_mcpu`, and the field `resources['req_cpu']`. Memory also has many forms, but it doesn't parallel all of them, for example whereas cpu does not have a variable `req_cpu`, memory does have `req_memory`.
- Inconsistent or non-analogue use of helper functions. For example, storage and cpu have "parse_X_in_UNITS" functions but for memory we have an `if` and two different functions.
- We don't always check for `None`, for example `parse_memory_in_bytes` can return `None`.
- Relatedly, we should add type annotations to protect us from the aforementioned error.
- I found it really hard to follow `select_inst_coll`. I think this is because it takes many arguments and it does not clearly distinguish between the two case: pool and private. I think that confusion comes from `front_end.py` which also does not make a clear distinction between pool and private. For me, the repeated conditionals were a symptom that we were not clearly distinguishing between the cases.
- I find implicit representations of distinct cases confusing. For example, the user specifies one of four cases for memory: `highmem`, `lowmem`, `standard`, or an explicit amount. We blur that into an explicit amount, an optionally requested number of cores, and an optional machine type. This creates two problems:
  - front_end.py is now both doing simple string parsing *and* machine selection (so as to convert a memory ratio to an explicit memory request)
  - InstanceCollectionConfigs has to recover this distinction from the memory amount, cores, and machine type. I believe this is how you ended up with `from_cost` and `from_machine_type`, which I found confusing.

In typical Dan-fashion, in order to understand what I found confusing, I had to manipulate the code myself. This is what I produced: https://github.com/jigold/hail/compare/expose-private-worker...danking:expose-private-worker. What do you think of incorporating those changes?